### PR TITLE
rfc: User IDs For Internal User Information

### PR DESCRIPTION
**Pull request by Fenil-P**
Release note: None

Release justification: non-code changes, rfc only

### CODE REVIEW

**Review by knz — COMMENTED** [review]
I like this proposal a lot. Would be curious to see more details about the proposed migration.

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-:-MxiPb8q0xofS7EQ2DB3:bk3srr2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Fenil-P)

___
*[docs/RFCS/20220307_user_ids.md, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiNQdOBTq0j4qnJWPz:-MxiNQdOBTq0j4qnJWQ-:bllh8gv) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L5)):*
> ```Markdown
> - Start Date: 2022-03-07
> - Authors: Fenil Patel
> - RFC PR: (PR # after acceptance of initial draft)
> ```

you can update this field now.
___
*[docs/RFCS/20220307_user_ids.md, line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiNT678t_nL1d3PIDv:-MxiNT678t_nL1d3PIDw:bkstsbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L6)):*
> ```Markdown
> - Authors: Fenil Patel
> - RFC PR: (PR # after acceptance of initial draft)
> - Cockroach Issue: [76079](https://github.com/cockroachdb/cockroach/issues/76079)
> ```

nit: we usually include `#` in front of the number.
___
*[docs/RFCS/20220307_user_ids.md, line 10 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiNYdE59Vvr1dhPpQo:-MxiNYdE59Vvr1dhPpQp:bftbj9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L10)):*
> ```Markdown
> 
> # Summary
> Support stable IDs to key internal user information instead of username. Requires a long-running migration to give 
> ```

Please also include a motivating sentence in the summary.
___
*[docs/RFCS/20220307_user_ids.md, line 10 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiNqCtByVvEJYYAMKw:-MxiNqCtByVvEJYYAMKx:b266ple) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L10)):*
> ```Markdown
> 
> # Summary
> Support stable IDs to key internal user information instead of username. Requires a long-running migration to give 
> ```

Please also say in the summary that those IDs will be UUIDs.
___
*[docs/RFCS/20220307_user_ids.md, line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiNWDu7lMN790YcYlD:-MxiNWDu7lMN790YcYlE:b-llhk1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L11)):*
> ```Markdown
> # Summary
> Support stable IDs to key internal user information instead of username. Requires a long-running migration to give 
> existing users IDs
> ```

nit: missing period at end of sentence.
___
*[docs/RFCS/20220307_user_ids.md, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiNemVBKs4paADjTj6:-MxiNemVBKs4paADjTj7:b-10v841) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L16)):*
> ```Markdown
> 
> Currently, all internal per-user information is keyed by the username. This is not ideal because it makes it difficult 
> to rename a user without breaking things. It's also not ideal because usernames can be considered personal data, but we 
> ```

you could provide an example of "thing" that breaks.
___
*[docs/RFCS/20220307_user_ids.md, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiNxPG3hRO7c8LlMPf:-MxiNxPG3hRO7c8LlMPg:b-9ib2) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L28)):*
> ```Markdown
> # Technical design
> ### Adding ID columns to tables with username columns
> The ids will be randomly generated UUID when creating a new user during insert into system.users. The SQL syntax for all 
> ```

What is "the SQL syntax for tables"?
___
*[docs/RFCS/20220307_user_ids.md, line 81 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiO2xk6UY9kDP3qqYU:-MxiO2xk6UY9kDP3qqYV:b1ig60) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L81)):*
> ```Markdown
> hashedPassword | BYTES     |    true     | NULL              |                       | {primary}                       |   false
> isRole         | BOOL      |    false    | false             |                       | {primary}                       |   false
> id             | UUID      |    false    | gen_random_uuid() |                       | {primary,users_username_id_idx} |   false
> ```

NB: column names based off a generic word (in this case, "id") should qualify this word to become more specific.

What's why we don't use just `name` and say `username` instead (which arguably would be better called `user_name`).
Ditto `database_id` instead of just `id`, etc.

So, here and throughout: `user_id`.
___
*[docs/RFCS/20220307_user_ids.md, line 103 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiOKNQ5aPLz0AXTisf:-MxiOKNQ5aPLz0AXTisg:b3wotcp) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L103)):*
> ```Markdown
> member      | STRING    |    false    | NULL           |                       | {primary,role_members_member_idx,role_members_role_idx} |   false
> isAdmin     | BOOL      |    false    | NULL           |                       | {primary}                                               |   false
> roleId      | UUID      |    false    | NULL           |                       | {primary,role_members_member_idx,role_members_role_idx} |   false
> ```

`role_id`

`member_id`

let's avoid capitals in new column names. `isAdmin`, `isRole` and `hashedPasswords` are regrettable historical accidents, no need to perpetrate them again.
___
*[docs/RFCS/20220307_user_ids.md, line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiOeam3a_9T4RnPPal:-MxiOeam3a_9T4RnPPam:b-f26z5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L122)):*
> ```Markdown
> ### Type of User ID
> - **UUID** allows for randomly generated IDs during insert using `gen_random_uuid()` as default value for the column.
>   #### Pros:
> ```

This breaks the formatting in the rendered page. Please use markdown formatting that makes sense.
___
*[docs/RFCS/20220307_user_ids.md, line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiOlAa9wfSATbDg4HK:-MxiOlAa9wfSATbDg4HL:btt7yd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L123)):*
> ```Markdown
> - **UUID** allows for randomly generated IDs during insert using `gen_random_uuid()` as default value for the column.
>   #### Pros:
>   - It requires no collision handling 
> ```

nit: here and below, make the bullet points full sentences, with a period at the end.
___
*[docs/RFCS/20220307_user_ids.md, line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiOp1C5WGhu_CmtQUt:-MxiOp1C5WGhu_CmtQUu:b-t2hi4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L126)):*
> ```Markdown
>   - Adding a new column with default value makes it easy to migrate older versions
>   #### Cons:
>   - Users can't make use of ids themselves in other statements as they are hard to remember
> ```

I don't understand this point: why would users would ever need to manipulate things by ID in _interactive_ sessions? Why is it important for them to remember anything?
___
*[docs/RFCS/20220307_user_ids.md, line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiOwZT440LQURNDVv4:-MxiOwZT440LQURNDVv5:b227f33) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L129)):*
> ```Markdown
> 
> 
> - **INT** is closer to what [postgres does](https://www.postgresql.org/docs/8.0/sql-createuser.html) with uid when creating a user. This is useful when allowing a user to specify the 
> ```

nit: wrap text at 80-100 columns.
___
*[docs/RFCS/20220307_user_ids.md, line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiP19591fAmUjrcTBR:-MxiP19591fAmUjrcTBS:bbf9aox) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L132)):*
> ```Markdown
> uid in future updates. For existing users, a hash function could be used to generated their ID. We can also take an approach similar to postgres and have the new ID be the highest ID + 1
>   #### Pros:
>   - Matches postgres use of UID (see example below)
> ```

I think you're underselling the utility of doing user IDs using a (pseudo-)sequence that fits integers.

The ability to generate a pg-compatible `pg_user` virtual table, with a useful `usesysid` column, may be more important than we think. Have you discussed this with our AppRel experts?
___
*[docs/RFCS/20220307_user_ids.md, line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiPNUu3Ozl7GD2WlQm:-MxiPNUu3Ozl7GD2WlQn:b4cgy51) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L174)):*
> ```Markdown
> An ID field will be added to UserPriviliges object. The Privilege Descriptors store user privilege objects in an array 
> in sorted order based on usernames currently. We’ll need to refactor the code such that the privileges array is sorted 
> by IDs instead.
> ```

1) why is the sorting important?

2) can we perhaps add a 2nd field alongside the usernames, during the migration period?
___
*[docs/RFCS/20220307_user_ids.md, line 195 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiPXTH18seXx5lXYBW:-MxiPXTIDcnRKvBf711w:b23k8m3) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L195)):*
> ```Markdown
> ### Long-running migration
> We need to add the new system table columns to existing instances and migrate all the existing users and generate their 
> corresponding IDs.The migration will need to update the UserPriviliges to contain the new user IDs.
> ```

nit: missing space after period.

**Review by RichardJCai — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77453)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Fenil-P and @knz)

___
*[docs/RFCS/20220307_user_ids.md, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiNemVBKs4paADjTj6:-MxjcHgD5Es9-N2gatiE:b-7ddeqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L16)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

you could provide an example of "thing" that breaks.
</blockquote></details>

I think breaking things might not be the right phrasing but rather it makes it difficult to support an operation like renaming a username difficult.
___
*[docs/RFCS/20220307_user_ids.md, line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiOp1C5WGhu_CmtQUt:-MxjchfSDJUNwhgupigY:b-oh76kf) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L126)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't understand this point: why would users would ever need to manipulate things by ID in _interactive_ sessions? Why is it important for them to remember anything?
</blockquote></details>

+1 
users shouldn't have to remember / use their internal ID.
___
*[docs/RFCS/20220307_user_ids.md, line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxjdV7Q2SnpNWssHTSu:-MxjdV7Q2SnpNWssHTSv:b-xrww6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L135)):*
> ```Markdown
>   - User can refer to them and possibly use in other statements
>   #### Cons:
>   - Migration will need to atomically add a NOT NULL column and fill in all the ids
> ```

We'd have to atomically add a NOT NULL column in the case of using UUID as well. The difference is we have to come up with another function to generate the ID here instead of gen_random_uuid. The con we have to come up with this function and the function will likely be less efficient.
___
*[docs/RFCS/20220307_user_ids.md, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxjdxPg7uEby-_5xsND:-MxjdxPg7uEby-_5xsNE:b-xago8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L198)):*
> ```Markdown
> 
> During the migration, we will need to update the existing Privilege descriptors and resort the UserPriviliges arrays by 
> ID in order to support the lookups by user IDs.
> ```

We also have to update default privileges.

I'd note it here that we have to update every field where a username is stored to also store an ID and then note that the only places right now are the various system tables, user privileges and default privileges.

**Review by RichardJCai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77453)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Fenil-P and @knz)

___
*[docs/RFCS/20220307_user_ids.md, line 126 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiOp1C5WGhu_CmtQUt:-MxjvhOeBMZcZanbOGgA:b-gfvd8x) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L126)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

+1 
users shouldn't have to remember / use their internal ID.
</blockquote></details>

Actually scratch the use part, they might want to use it but they should access it programmatically from the `pg_user` table if so.

**Review by Fenil-P — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77453)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Fenil-P, @knz, and @RichardJCai)

___
*[docs/RFCS/20220307_user_ids.md, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiNQdOBTq0j4qnJWPz:-MxogCXL3J16PRWbvFAX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L5)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

you can update this field now.
</blockquote></details>

Done.
___
*[docs/RFCS/20220307_user_ids.md, line 10 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiNYdE59Vvr1dhPpQo:-MxogEsJ3YQ1unLcBSXP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L10)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please also include a motivating sentence in the summary.
</blockquote></details>

Done.
___
*[docs/RFCS/20220307_user_ids.md, line 10 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiNqCtByVvEJYYAMKw:-MxogFuD9SSF7RME8OGf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L10)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please also say in the summary that those IDs will be UUIDs.
</blockquote></details>

Done.
___
*[docs/RFCS/20220307_user_ids.md, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiNemVBKs4paADjTj6:-MxogHurAOt9OlIgq3dE:b-bu91xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L16)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I think breaking things might not be the right phrasing but rather it makes it difficult to support an operation like renaming a username difficult.
</blockquote></details>

Updated
___
*[docs/RFCS/20220307_user_ids.md, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiNxPG3hRO7c8LlMPf:-MxogJgQ5szELrLTV70i:b-h1a2ru) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L28)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

What is "the SQL syntax for tables"?
</blockquote></details>

Updated, I was referring to inserts, select and etc...
___
*[docs/RFCS/20220307_user_ids.md, line 81 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiO2xk6UY9kDP3qqYU:-MxogKvZ3_RHKcjL_ouK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L81)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

NB: column names based off a generic word (in this case, "id") should qualify this word to become more specific.

What's why we don't use just `name` and say `username` instead (which arguably would be better called `user_name`).
Ditto `database_id` instead of just `id`, etc.

So, here and throughout: `user_id`.
</blockquote></details>

Done.
___
*[docs/RFCS/20220307_user_ids.md, line 103 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiOKNQ5aPLz0AXTisf:-MxogLm--FQClC9SiM6a:bwygxsk) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L103)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`role_id`

`member_id`

let's avoid capitals in new column names. `isAdmin`, `isRole` and `hashedPasswords` are regrettable historical accidents, no need to perpetrate them again.
</blockquote></details>

Noted!
___
*[docs/RFCS/20220307_user_ids.md, line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiOeam3a_9T4RnPPal:-Mxp8iLJB-46tL_91h3O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L122)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This breaks the formatting in the rendered page. Please use markdown formatting that makes sense.
</blockquote></details>

Done.
___
*[docs/RFCS/20220307_user_ids.md, line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxjdV7Q2SnpNWssHTSu:-Mxp8UgsDK1mpJycGurx:b-bu91xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L135)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

We'd have to atomically add a NOT NULL column in the case of using UUID as well. The difference is we have to come up with another function to generate the ID here instead of gen_random_uuid. The con we have to come up with this function and the function will likely be less efficient.
</blockquote></details>

Updated
___
*[docs/RFCS/20220307_user_ids.md, line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxiPNUu3Ozl7GD2WlQm:-Mxp8E6G4BkCRUaf9vUJ:b5erigc) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L174)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) why is the sorting important?

2) can we perhaps add a 2nd field alongside the usernames, during the migration period?
</blockquote></details>

Added two options for lookup implementation, please take a look and let me know what you think!
___
*[docs/RFCS/20220307_user_ids.md, line 198 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-MxjdxPg7uEby-_5xsND:-Mxp8AF9At1Qjt1PBhsR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/27436d8b5fe06a056c7d2b7b3b68cb0524e6818c/docs/RFCS/20220307_user_ids.md#L198)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

We also have to update default privileges.

I'd note it here that we have to update every field where a username is stored to also store an ID and then note that the only places right now are the various system tables, user privileges and default privileges.
</blockquote></details>

Done.

File: docs/RFCS/20220307_user_ids.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+# High-level Summary
+The idea is to convert code dependencies of username and lookups to use the IDs instead in 3 steps. First adding the ID
+columns to system tables and update related syntax statements. Second, updating the user privileges objects and lookup
+based on usernames. Finally, migrating the existing instances of Cockroach.
+
+
+# Technical design
+### Adding ID columns to tables with username columns
+The ids will be randomly generated UUID when creating a new user during insert into system.users. SQL statements
```
Comment (RichardJCai) [inline docs/RFCS/20220307_user_ids.md:33]: IDs are oids.

File: docs/RFCS/20220307_user_ids.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+#### After Change:
+```
+demo@127.0.0.1:26257/movr> \d system.users
+column_name   | data_type | is_nullable |  column_default   | generation_expression |             indices             | is_hidden
+-----------------+-----------+-------------+-------------------+-----------------------+---------------------------------+------------
+username       | STRING    |    false    | NULL              |                       | {primary,users_username_id_idx} |   false
+hashedPassword | BYTES     |    true     | NULL              |                       | {primary}                       |   false
+isRole         | BOOL      |    false    | false             |                       | {primary}                       |   false
+user_id        | UUID      |    false    | gen_random_uuid() |                       | {primary,users_username_id_idx} |   false
```
Comment (RichardJCai) [inline docs/RFCS/20220307_user_ids.md:86]: oid for this and below

File: docs/RFCS/20220307_user_ids.md [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+ab       |    16389 | f           | f        | f       | f            | ******** |          |
+(6 rows)
+```
+
+***Require more discussion on this**
+
+
+
+### Updating User Privileges and Lookups
+An ID field will be added to UserPrivileges and any objects that contain a username. The Privilege Descriptors store
```
Comment (RichardJCai) [inline docs/RFCS/20220307_user_ids.md:193]: Note that in 23.1, we can remove the username field from privileges.

**Review by RichardJCai — APPROVED** [review]
Let's merge this RFC

**Review by Fenil-P — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77453)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Fenil-P, @knz, and @RichardJCai)

___
*[`docs/RFCS/20220307_user_ids.md` line 33 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-Mxa4YWD5BNMNRptAjkI-r3-33:-N0WmS4Q1lZSC10AdivV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b29b81efde3036ed22a0af788ea375a74e4d9a6/docs/RFCS/20220307_user_ids.md#L33)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

IDs are oids.
</blockquote></details>

Done.
___
*[`docs/RFCS/20220307_user_ids.md` line 86 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77453#-Mxa4YWD5BNMNRptAjkI-r3-86:-N0WmT3x61rgenq9eKf5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b29b81efde3036ed22a0af788ea375a74e4d9a6/docs/RFCS/20220307_user_ids.md#L86)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

oid for this and below
</blockquote></details>

Done.

**Pull request merged**